### PR TITLE
Make the quorum generator & Validator routines @safe and nothrow

### PR DESCRIPTION
**Pull request by AndrejMitrovic**

### CODE REVIEW

**Comment by Geod24** [comment]
> For whatever reason the autogenerated opAssign [...]

Maybe because the dtor is `@system` ?

File: source/agora/consensus/protocol/Nominator.d [hunk 1]
```diff
@@ -122,23 +122,23 @@ extern(D):
...
     {
         assert(!this.is_nominating);
-        this.quorum_set.clear();
+        () @trusted { this.quorum_set.clear(); }();
 
         // store the list of other node's quorum hashes
         foreach (qc; other_quorums)
         {
-            auto quorum_set = verifyBuildSCPConfig(qc);
+            auto quorum_set = buildSCPConfig(qc);
```
Comment (linked0) [inline source/agora/consensus/protocol/Nominator.d:133]: Would it be less confusing to use a different name from the `this.quorum_set`?

File: source/agora/consensus/protocol/Nominator.d [hunk 2]
```diff
@@ -122,23 +122,23 @@ extern(D):
...
             auto shared_set = makeSharedSCPQuorumSet(quorum_set);
             this.quorum_set[hashFull(quorum_set)] = shared_set;
         }
 
         // set up our own quorum
-        auto quorum_set = verifyBuildSCPConfig(quorum);
-        this.scp.updateLocalQuorumSet(quorum_set);
-        auto shared_set = makeSharedSCPQuorumSet(this.scp.getLocalQuorumSet());
+        auto quorum_set = buildSCPConfig(quorum);
```
Comment (linked0) [inline source/agora/consensus/protocol/Nominator.d:139]: Same question for the name for the variable.

File: source/agora/consensus/protocol/Nominator.d (hunk 1 above)
Comment (AndrejMitrovic) [reply source/agora/consensus/protocol/Nominator.d:133]: Yes, in fact the field is badly named. I will add another commit.

**Comment by AndrejMitrovic** [comment]
Added another commit renaming the field as suggested.

**Comment by AndrejMitrovic** [comment]
> Maybe because the dtor is @system ?

Yes that was it. I wish the diagnostic was better.

**Comment by AndrejMitrovic** [comment]
https://issues.dlang.org/show_bug.cgi?id=21043

File: source/scpd/scp/QuorumSetUtils.d [hunk 3]
```diff
@@ -22,4 +22,5 @@ bool isQuorumSetSane(ref const SCPQuorumSet qSet, bool extraChecks,
     const(char)** reason = null);
 
 // normalize the quorum set, optionally removing idToRemove
-void normalizeQSet(ref SCPQuorumSet qSet, const(NodeID)* idToRemove = null);
+void normalizeQSet(ref SCPQuorumSet qSet, const(NodeID)* idToRemove = null)
+    nothrow @nogc;
```
Comment (MichaelKim20) [inline source/scpd/scp/QuorumSetUtils.d:26]: There's something I don't understand, so let me ask you.
Use 'try catch' at C++ to modify it.
Or should we pack this function once more and make an exception there?

File: source/scpd/types/Utils.d [hunk 4]
```diff
@@ -19,7 +19,7 @@ import scpd.types.Stellar_types;
 import scpd.types.XDRBase;
 
 extern(C++) public shared_ptr!SCPQuorumSet makeSharedSCPQuorumSet (
-    ref const(SCPQuorumSet));
+    ref const(SCPQuorumSet)) nothrow @nogc @safe;
```
Comment (MichaelKim20) [inline source/scpd/types/Utils.d:22]: I have the same question as above.
There's something I don't understand, so let me ask you.
Add exception processing at C++.
Or should we pack this function once more and make an exception there?

**Review by MichaelKim20 — COMMENTED** [review]
Very good. I have two questions.

File: source/scpd/scp/QuorumSetUtils.d (hunk 3 above)
Comment (AndrejMitrovic) [reply source/scpd/scp/QuorumSetUtils.d:26]: This routine does not throw any exceptions.

File: source/scpd/types/Utils.d (hunk 4 above)
Comment (AndrejMitrovic) [reply source/scpd/types/Utils.d:22]: The routine does not throw. We're passing a reference and it's calling `std::make_shared`, it won't fail.

File: source/scpd/scp/QuorumSetUtils.d (hunk 3 above)
Comment (MichaelKim20) [reply source/scpd/scp/QuorumSetUtils.d:26]: I see.

**Comment by Geod24** [comment]
Still a draft ?

**Comment by AndrejMitrovic** [comment]
No, just infinite wait bug.

**Comment by AndrejMitrovic** [comment]
I'll rebase.

**Review by MichaelKim20 — APPROVED** [review]
LGTM

**Comment by AndrejMitrovic** [comment]
@Geod24 do you want to look into it more, or it looks good to you?

**Comment by AndrejMitrovic** [comment]
If you want me to squash some stuff let me know. 👋

File: source/agora/consensus/Quorum.d [hunk 5]
```diff
@@ -87,11 +87,18 @@ public QuorumConfig buildQuorumConfig ( const ref PublicKey key,
     auto RNG_gen = getGenerator(key);
     auto stake_amounts = stakes.map!(stake => stake.amount.integral);
 
+    auto assumeNothrow (T)(lazy T exp) nothrow
```
Comment (Geod24) [inline source/agora/consensus/Quorum.d]:
```suggestion
    static auto assumeNothrow (T)(lazy T exp) nothrow
```

File: source/agora/consensus/Quorum.d [hunk 6]
```diff
@@ -70,7 +70,7 @@ private enum double THRESHOLD = 0.80;
 *******************************************************************************/
 
 public QuorumConfig buildQuorumConfig ( const ref PublicKey key,
-    in Hash[] utxo_keys, UTXOFinder finder )
+    in Hash[] utxo_keys, UTXOFinder finder ) @safe nothrow
```
Comment (Geod24) [inline source/agora/consensus/Quorum.d]:
```suggestion
    in Hash[] utxo_keys, scope UTXOFinder finder ) @safe nothrow
```

**Review by Geod24 — APPROVED** [review]
Just two nits, otherwise LGTM. Merge at will. I'd say squash the 3 commits that add `@safe` here and there if you want, but I'm okay either way.

**Pull request merged**